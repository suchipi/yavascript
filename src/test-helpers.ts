import { spawn, Options as SpawnOptions } from "first-base";
import child_process from "child_process";
import path from "path";
import print from "@suchipi/print";
import * as inspectOptions from "./inspect-options";

export const TMP = child_process
  .execSync("realpath /tmp", { encoding: "utf-8" })
  .trim();

export const rootDir = (...parts: Array<string>) =>
  path.resolve(__dirname, "..", ...parts);

function getBinaryPath(platform: string, arch: string) {
  if (platform === "win32") {
    return path.resolve(__dirname, "..", "bin", "windows", "yavascript.exe");
  } else if (platform === "darwin") {
    if (arch.startsWith("arm")) {
      return path.resolve(__dirname, "..", "bin", "darwin-arm", "yavascript");
    } else {
      return path.resolve(__dirname, "..", "bin", "darwin", "yavascript");
    }
  } else if (platform === "linux") {
    return path.resolve(__dirname, "..", "bin", "linux", "yavascript");
  } else {
    throw new Error("Unsupported platform: " + platform);
  }
}

var binaryPath = getBinaryPath(
  process.platform,
  process.env.YS_TESTS_ARCH || process.arch
);

export { binaryPath, getBinaryPath };

export type EvaluateResult = {
  stdout: string;
  stderr: string;
  code: number | null;
  error: boolean;
};

export async function evaluate(
  code: string,
  options: SpawnOptions = {}
): Promise<EvaluateResult> {
  const runContext = spawn(binaryPath, ["-e", code], options);
  await runContext.completion;
  return runContext.result;
}

export function inspect(value: any): string {
  // options to inspect here match what is given to console.log
  return print(value, inspectOptions.forPrint);
}

export function cleanOutput(input: string): string {
  return input
    .replace(/yavascript:\d+/g, "yavascript")
    .replace(/\/bin\/[^/]+\/yavascript/g, "/bin/.../yavascript")
    .replace(new RegExp(rootDir(), "g"), "<rootDir>");
}

export function cleanResult(input: EvaluateResult): EvaluateResult {
  return {
    ...input,
    stdout: cleanOutput(input.stdout),
    stderr: cleanOutput(input.stderr),
  };
}
