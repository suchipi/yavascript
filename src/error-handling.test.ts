import { evaluate } from "./test-helpers";

test("prints thrown errors to stderr", async () => {
  const result = await evaluate(`blahhhh`);
  expect(result).toMatchInlineSnapshot(`
    {
      "code": 1,
      "error": false,
      "stderr": "ReferenceError: 'blahhhh' is not defined
      at <eval>
    ",
      "stdout": "",
    }
  `);
});

test("prints thrown non-errors to stderr", async () => {
  const result = await evaluate(`throw "nope"`);
  expect(result).toMatchInlineSnapshot(`
    {
      "code": 1,
      "error": false,
      "stderr": "Non-error value was thrown: "nope"
    ",
      "stdout": "",
    }
  `);
});

test("prints extra error properties to stderr", async () => {
  const result = await evaluate(
    `e = new Error('hi'); e.something = true; e.somethingElse = false; throw e;`
  );
  expect(result).toMatchInlineSnapshot(`
    {
      "code": 1,
      "error": false,
      "stderr": "Error: hi
      at <eval>
      something: true
      somethingElse: false
    }
    ",
      "stdout": "",
    }
  `);
});
