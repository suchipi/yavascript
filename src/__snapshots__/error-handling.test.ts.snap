// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prints extra error properties to stderr 1`] = `
{
  "code": 1,
  "error": false,
  "stderr": "Error: hi
  at <eval> (<evalScript>) {
  something: true
  somethingElse: false
}
",
  "stdout": "",
}
`;

exports[`prints thrown errors to stderr 1`] = `
{
  "code": 1,
  "error": false,
  "stderr": "ReferenceError: 'blahhhh' is not defined
  at <eval> (<evalScript>)
",
  "stdout": "",
}
`;

exports[`prints thrown non-errors to stderr 1`] = `
{
  "code": 1,
  "error": false,
  "stderr": "Non-error value was thrown: "nope"
",
  "stdout": "",
}
`;
